Add lists filtering

saveriomiroddi opened the pull request:
Give the user the ability to filter the current list entries using alphanumeric characters and a few other characters, the same way Github multiple choice widgets work.

Sample instantiation and output:

```ruby
prompt.select("Select drinks?", %w(vodka beer wine whisky bourbon), filter: true)
```

```
    Select drinks? (Pattern: "k")
    ‣ vodka
      whisky
```

This option, activated via `filter: true` is incompatible with the enumeration, in order to avoid ambiguity when typing numbers.

An alternative code design is to make `@choices` mutable, and store the original choices in a separate variable; this minimizes the changes, at the cost of increased mutability.

Closes #71.

saveriomiroddi commented:
I'm unsure how to proceed with the linting warnings about line modified, which were already in violation of the guidelines.

piotrmurach commented:
This is looking really good and I'm excited about this feature! However, I think there is few things that will need to be sorted out before merge:
*  I would really encourage your to look into providing tests for both the `select` & `multi_select` prompts. 
* I'm not super sold on the `pattern_matching` option name. Naming is hard, but I think it really benefits the api when it's intuitive. How about calling it `filter` since the intention behind this is to filter options out?
* Consider expanding the help to say `press key to filter options` etc...

Just few ideas quickly reviewing your PR. I wouldn't worry too much about the linting errors, they are there for guidance rather than strict execution 😄

saveriomiroddi commented:
So! I've addressed all the points mentioned, assuming I've understood well (I would have never released this without unit testing :wink: ).

As a bonus, I've also bound the `Canc` key, so that the filter can be erased with a single keystroke. I've originally tried to bind `Esc` via `def keyescape`, but I don't know why, the event wasn't sent or listened, and the method was not executed.

A note:

> Consider expanding the help to say press key to filter options etc...

I hope I understood this well (you can check the UTs). I don't know if I should add also mention of backspace and canc, as the help line is getting quite long. Just let me know your thoughts, and I'll apply them.

The unit testing it's arguably simplistic, but it covers a reasonable amount of cases.

saveriomiroddi commented:
OK, once we finish retouching the code (in case there are other iteration), I'll update the README.

saveriomiroddi commented:
@piotrmurach Can you restart the travis build? It seems it got stuck - the hosts didn't boot: https://travis-ci.org/piotrmurach/tty-prompt/builds/329680132

saveriomiroddi commented:
@piotrmurach should adding `def keyescape(*)` receive the event (in Multi/List)? I would have actually preferred to bind Esc rather than Canc to clear the filter, however, since what I tried doesn't work, I wonder if I did something wrong, or if it's a bug.

saveriomiroddi commented:
I've opened an issue on `tty-reader`, as I think it's a bug (or a limitation of the implementation): https://github.com/piotrmurach/tty-reader/issues/6.

saveriomiroddi commented:
I've realized that there is a suite of spec specific for `multi_select`. I'll implement tests also for that one.

In the meanwhile, a design question. What do you think of case insensitive search? There are two approaches:

1. make filtering case insensitive
2. enable case insensitivity an option

if you think 2. would be the better, there are different way to specify the option, either by overloading the `filter` option (ie. by allowing both `filter: true`,  `filter: ignorecase`) or by adding a separate option (ie. by allowing `filter: true` and `filter_case`)

piotrmurach commented on lib/tty/prompt/list.rb:
```diff
@@ -319,6 +373,30 @@ def render_footer
+      # @return [Choice]
+      #
+      # @api private
+      def visible_choices
```
My preference would be to name it simply `choices` as the fact that they are visible is just a presentation concept. Here we're dealing with data transformation and what happens with in the terminal display is of no concern. What follows, I would also make this method public as exposing it may lead to some interesting use cases where people can inspect their current choices.

saveriomiroddi replied:
There is a already a [:choices method](https://github.com/saveriomiroddi/tty-prompt-dev/blob/a16277fb21561502bc5ab18ce7ba16eb80bd1567/lib/tty/prompt/list.rb#L160) :smile:

piotrmurach replied:
Hmmm...with few methods I apply a style, in which the `choices` could act as reader and writer of the values. When no value is provided it returns the filtered/all choices:

```ruby
def choices(values = (not_set = true))
  if not_set
     filtered/all choices....
  else
     current behaviour....
  end
end
```
How do you feel about it?

saveriomiroddi replied:
He! I've never seen it, so it seems strange/unusual. It's a trivial change though, so I've just applied it.

Let me know if you think unit testing is required, since in `select_spec.rb` there is none (at least, based on what I see :smile:), but both getter and setter version are tested extensively (functionally).

piotrmurach commented on lib/tty/prompt/list.rb:
```diff
@@ -13,10 +15,15 @@ class Prompt
 
       PAGE_HELP = '(Move up or down to reveal more choices)'
 
+      KEY_CANC = "\e[3~".freeze
```
Ideally, any low level terminal codes should be handled by `tty-reader` and we should rely on its api. Partly, because this library needs to work on different systems including Windows, which it is currently failing on, and keys have very different codes there. I've noticed your ticket in `tty-reader` repository, thank you!

piotrmurach commented on lib/tty/prompt/list.rb:
```diff
@@ -111,7 +121,16 @@ def help(value = (not_set = true))
+        tokens = if enumerate?
+                   [" or number (1-#{visible_choices.size})", ""]
+                 elsif @choices_filter
+                   ["", ", and alphanumeric/underscore characters to filter"]
```
The wording could be a bit more user friendly. Though it is technically correct, I think it is important to remember that CLI users may not necessarily be technically savvy.  I propose something akin to `and press letter keys to filter` or similar.

saveriomiroddi replied:
:+1: I've changed `alphanumeric/underscore characters` to `letter keys`, so that it now reads `press Enter to select, and letter keys to filter`.

piotrmurach commented on spec/unit/select_spec.rb:
```diff
@@ -400,7 +400,136 @@ def output_helper(prompt, choices, active, options = {})
     }.to raise_error(TTY::Prompt::ConfigurationError, /`10` out of range \(1 - 3\)/)
   end
+
+  context "with filter" do
```
I would encourage you to use `output_helper` to make the specs more consistent. Especially when it comes down to symbols for select menus as these currently fail on Windows.

piotrmurach commented on spec/unit/multi_select_spec.rb:
```diff
@@ -375,4 +375,42 @@ def output_helper(prompt, choices, active, selected, options = {})
     )
   end
+
+  context "with filter" do
```
I would encourage you to make specs more consistent with already present either by using the `output_helper` or manually creating the output string. Currently these fail on Windows.

saveriomiroddi commented:
Excellent points. I'll apply them over the next couple of days.

saveriomiroddi commented:
@piotrmurach ok, I've addressed everything :smile: The only open point is the `[visible_]choices` method.

piotrmurach commented on README.md:
````diff
@@ -778,6 +779,37 @@ prompt.multi_select("Select drinks?", choices, echo: false)
 # ‣ ⬢ 5) bourbon
 ```
 
+### 2.6.2.1 echo
````
Probably copy&paste error?

saveriomiroddi replied:
Yikes! Addressed.

piotrmurach commented on README.md:
```diff
@@ -778,6 +779,37 @@ prompt.multi_select("Select drinks?", choices, echo: false)
+### 2.6.2.1 echo
+
+To activate dynamic list filtering on letter/number typing, use the :filter option:
+header use the :echo option:
```
Remnant from copy & paste?

saveriomiroddi replied:
Addressed as well.

piotrmurach commented:
I think the search should be case insentitive by default, and I do struggle to think why anyone would need to have case sensitive search. Let's go with an option 1, however, if you feel strongly about option 2 then I would suggest `ignore_case: true` as a default option.

All in all, this is a great pice of work Saverio and I do appreciate it! We're nearly there, just a few extra changes and we can merge this PR!

saveriomiroddi commented:
> I think the search should be case insentitive by default, and I do struggle to think why anyone would need to have case sensitive search. Let's go with an option 1

That's great :smile:Search is case insensitive

saveriomiroddi commented:
@piotrmurach everything is addressed. let me know if I should squash some/the commits. all the commits are independent and stable (no UT failures) pieces of work (although the review commits are improvements of the previous).
I've fixed up the README correction in the existing README commit.

saveriomiroddi commented:
by the way, Hound is really noisy - it's really hard to follow the discussions, as it added over a hundred commits :scream:
Not sure if there is an easy way to solve this. Possibly, if you add linting via Rubocop, you can get away with automatic cleanup of most/many of them.

piotrmurach commented:
Hound service is annoying, however it sometimes picks on things so that people clean up their PRs tiny bit. I may consider switching it off as it annoys me as well! 

If it's not too much hassle I would prefer a squashed commit. Thanks!

saveriomiroddi commented:
@piotrmurach squash served!

piotrmurach commented:
Hey Saverio,

Thank you for your great contribution! I really enjoyed collaborating with you and I really appreciate you for taking my feedback and acting on it. I've been travelling recently and have some big travels ahead of me but I will endeavour to release new version later this week! Please bear with me.
